web/client_service_indexeddb.ts

```diff
@@ -0,0 +1,111 @@
+import Dexie, {Table} from 'dexie';
+import {KEYUTIL, b64utohex} from 'jsrsasign';
+import {Properties, CLIENT_PROPERTIES_FIELD} from './common/client_config';
+import {ClientService} from './client_service';
+import {isNonEmptyString, isInteger} from './common/polyfill/is';
+
+const KEYSTORE_DATABASE_NAME = 'scalar';
+
+class ScalarIndexedDb extends Dexie {
+  keystore!: Table<{id: string; key: CryptoKey}, string>;
+  certstore!: Table<{id: string; cert: string}, string>;
+
+  constructor() {
+    super(KEYSTORE_DATABASE_NAME);
+    this.version(1).stores({keystore: 'id', certstore: 'id'});
+  }
+}
+
+export class ClientServiceWithIndexedDb extends ClientService {
+  static async create(properties: Properties) {
+    const holderId = properties[CLIENT_PROPERTIES_FIELD.CERT_HOLDER_ID];
+    const certVersion = properties[CLIENT_PROPERTIES_FIELD.CERT_VERSION];
```

Comment:
I guess HMAC will be supported in JavaScript Client SDK in the future, not now. However, it might be nice if we consider how we can handle both digital signature and HMAC in a smart way even at this moment. I think the abstraction of DL (e.g., entity ID and key version or something like that) might be a good reference. I'm not sure a similar abstraction can be applied here for IndexedDB, but I just want to know what you think about this idea.

Reply:
@jnmt 

Thank you for your question and reminder, and sorry for my late reply.

We have these properties in client.properties after 3.7

- scalar.dl.client.cert_holder_id
- scalar.dl.client.cert_version
- scalar.dl.client.cert_pem
- scalar.dl.client.private_key_pem
- scalar.dl.client.entity_id
- scalar.dl.client.entity.identity.digital_signature.cert_version
- scalar.dl.client.entity.identity.digital_signature.cert_pem
- scalar.dl.client.entity.identity.digital_signature.private_key_pem
- scalar.dl.client.entity.identity.hmac.secret_key
- scalar.dl.client.entity.identity.hmac.key_version

In this PR, we cache `scalar.dl.client.cert_pem` and `scalar.dl.client.private_key_pem` in indexedDB like

```
indexedDB: {
    keystore: {
        "{cert_holder_id}_{cert_verion}": "{CryptoKey from private_key_pem}"
    },
    certstore: {
        "{cert_holder_id}_{cert_verion}": "{cert_pem}"
    }
}
```

I only handled old properties in this PR before 3.8 because I need to release ScalarDL Node/Web SDKs for version 3.6 and 3.7.

I plan to create another PR regarding HMAC supports.

In that PR, I would add three more databases:

```
indexedDB: {
    digital_signature_key_store: {
        "{entity_id}_{digital_signature.cert_version}": "{CtyptoKey from digital_signature.private_key_pem}"
    },
    digitail_signature_cert_store: {
        "{entity_id}_{digital_signature.cert_version}": "{digital_signature.cert_pem}"
    },
    hmac_key_store: {
        "{entity_id}_{hmac.key_version}": "{hmac.secret_key}"
    }
}
```

And, in ClientServiceWithIndexedDB class, I won't try to determine to store old or new properties contents. I would just store all of them if they are configured in client.properties. On the other hand, I will try to load all contents from indexedDB (according to proper entity_id or holder_id) anyway.

I would try to make this class as simple.

So, I might remove the error probing like here
https://github.com/scalar-labs/scalardl-javascript-client-sdk/pull/35/files#diff-6f96d6ce91ab0023414402b48c1fb57df5a3c2aca77f2f96cb0dcd798d4ff7b8R66

I would just let the ClientConfig class to check if required properties are configured, and, also let ClientConfig determine what properties it should take if old and new properties exist at the same.

This is my idea. How do you think? Does it make sense to you?

Reply:
@supl Thank you for the detailed explanation. Now I think I almost understand what you think. I agree it's good to make this class just load props as is without caring about the difference between v3.6 style and the v3.7- style for simplicity.